kivy/uix/textinput.py

```diff
@@ -835,7 +856,186 @@ def _set_undo_redo_bkspc(self, ol_index, new_index, substring, from_undo):
         #reset redo when undo is appended to
         self._redo = []
 
-    def do_cursor_movement(self, action):
+    def _move_word_left(self, index=None):
+        pos = index or self.cursor_index()
+        text = self.text
+        pos -= 1
+
+        if pos == 0:
+            return 0, 0
+
+        ucase = string.ascii_uppercase
+        lcase = string.ascii_lowercase
+        ws = string.whitespace
+        punct = string.punctuation
+
+        mode = 'normal'
+
+        c = text[pos]
+        if c in ws:
+            mode = 'ws'
+        elif c == '_':
+            mode = 'us'
+        elif c in punct:
+            mode = 'punct'
+        elif c not in ucase:
+            mode = 'camel'
+
+        while 0 < pos:
+            lc = c
+            c = text[pos]
+            if c == '\n':
+                if lc not in ws:
+                    pos += 1
+                break
+            if mode in ('normal', 'camel') and c in ws:
+                pos += 1
+                break
+            if mode in ('normal', 'camel') and c in punct:
+                pos += 1
+                break
+            if mode == 'camel' and c in ucase:
+                break
+            if mode == 'punct' and (c == '_' or c not in punct):
+                pos += 1
+                break
+            if mode == 'us' and c != '_' and (c in punct or c in ws):
+                pos += 1
+                break
+
+            if mode == 'us' and c != '_':
+                mode = ('normal' if c in ucase
+                        else 'ws' if c in ws
+                        else 'camel')
+            elif mode == 'ws' and c not in ws:
+                mode = ('normal' if c in ucase
+                        else 'us' if c == '_'
+                        else 'punct' if c in punct
+                        else 'camel')
+
+            pos -= 1
+
+        return self.get_cursor_from_index(max(0, pos))
+
+    def _move_word_right(self, index=None):
+        pos = index or self.cursor_index()
+        text = self.text
+        pmax = len(text)
+
+        if pos == pmax:
+            return self.cursor
+
+        ucase = string.ascii_uppercase
+        lcase = string.ascii_lowercase
+        ws = string.whitespace
+        punct = string.punctuation
+
+        mode = 'normal'
+
+        c = text[pos]
+        if c in ws:
+            mode = 'ws'
+        elif c == '_':
+            mode = 'us'
+        elif c in punct:
+            mode = 'punct'
+        elif c in lcase:
+            mode = 'camel'
```

Comment:
isn't this the same code as above, should we not move the common bits to a common function and not duplicate?